torchdata/datapipes/iter/util/header.py

```diff
@@ -33,25 +33,29 @@ class HeaderIterDataPipe(IterDataPipe[T_co]):
         [0, 1, 2]
     """
 
-    def __init__(self, source_datapipe: IterDataPipe[T_co], limit: int = 10) -> None:
+    def __init__(self, source_datapipe: IterDataPipe[T_co], limit: Union[int, None] = 10) -> None:
         self.source_datapipe: IterDataPipe[T_co] = source_datapipe
-        self.limit: int = limit
+        self.limit: Union[int, None] = limit
 
     def __iter__(self) -> Iterator[T_co]:
         i: int = 0
         for value in self.source_datapipe:
             i += 1
-            if i <= self.limit:
+            if self.limit is None or i <= self.limit:
                 yield value
             else:
                 break
-        self.length = min(i, self.limit)  # We know length with certainty when we reach here
+        # We know length with certainty when we reach here
+        self.length = i if self.limit is None else min(i, self.limit)
```

Comment:
This can probably be simplified to using just `i` without a min.  As incrementing is done before yielding or breaking, the length is `i-1`.

The following test case should be added: use a generator (has no len function) with some yields, drain it and assert self.length afterwards.

Reply:
TBH, I think `self.length` should even be eliminated here for the reason that the `DataPipe` graph for each epoch might have different lengths as users can always alter the graph after one epoch ends.
We can simply rely on `limit` and `len(self.source_datapipe)` IMHO